src/handler.rs

```diff
@@ -7,69 +7,143 @@ use actix_service::{
 
-pub fn handler_service<F, T, R>(
-    handler: F,
+pub fn handler_service<H, T>(
+    handler: H,
 ) -> BoxServiceFactory<(), ServiceRequest, ServiceResponse, Error, ()>
 where
-    F: Handler<T, R>,
-    T: FromRequest,
-    R: Future,
-    R::Output: Responder,
+    H: for<'a> Handler<'a, T>,
+    T: for<'a> FromRequestX<'a>,
 {
     boxed::factory(fn_service(move |req: ServiceRequest| {
         let handler = handler.clone();
         async move {
             let (req, mut payload) = req.into_parts();
             let res = match T::from_request(&req, &mut payload).await {
                 Err(err) => HttpResponse::from_error(err),
-                Ok(data) => handler.call(data).await.respond_to(&req),
+                Ok(data) => handler.handle(data).await.respond_to(&req),
             };
             Ok(ServiceResponse::new(req, res))
         }
     }))
 }
 
+/// Same as [`std::ops::Fn`]
+pub trait FnX<Args> {
+    type Output;
+    fn call(&self, args: Args) -> Self::Output;
+}
+
 /// FromRequest trait impl for tuples
-macro_rules! factory_tuple ({ $($param:ident)* } => {
-    impl<Func, $($param,)* Res> Handler<($($param,)*), Res> for Func
-    where Func: Fn($($param),*) -> Res + Clone + 'static,
-          Res: Future,
-          Res::Output: Responder,
+macro_rules! fn_tuple ({ $($param:ident)* } => {
+    impl<Func, $($param,)* O> FnX<($($param,)*)> for Func
+    where Func: Fn($($param),*) -> O,
     {
+        type Output = O;
+
         #[allow(non_snake_case)]
-        fn call(&self, ($($param,)*): ($($param,)*)) -> Res {
+        fn call(&self, ($($param,)*): ($($param,)*)) -> O {
             (self)($($param,)*)
         }
     }
 });
 
-factory_tuple! {}
-factory_tuple! { A }
-factory_tuple! { A B }
-factory_tuple! { A B C }
-factory_tuple! { A B C D }
-factory_tuple! { A B C D E }
-factory_tuple! { A B C D E F }
-factory_tuple! { A B C D E F G }
-factory_tuple! { A B C D E F G H }
-factory_tuple! { A B C D E F G H I }
-factory_tuple! { A B C D E F G H I J }
-factory_tuple! { A B C D E F G H I J K }
-factory_tuple! { A B C D E F G H I J K L }
+fn_tuple! {}
+fn_tuple! { A }
+fn_tuple! { A B }
+fn_tuple! { A B C }
+fn_tuple! { A B C D }
+fn_tuple! { A B C D E }
+fn_tuple! { A B C D E F }
+fn_tuple! { A B C D E F G }
+fn_tuple! { A B C D E F G H }
+fn_tuple! { A B C D E F G H I }
+fn_tuple! { A B C D E F G H I J }
+fn_tuple! { A B C D E F G H I J K }
+fn_tuple! { A B C D E F G H I J K L }
+
+#[cfg(test)]
+mod test {
+    use serde::Deserialize;
+
+    use super::*;
+    use crate::{
+        dev::Service,
+        http::StatusCode,
+        test::{init_service, TestRequest},
+        web, App, HttpRequest, HttpResponse,
+    };
+
+    #[derive(Deserialize)]
+    struct Params<'a> {
+        name: &'a str,
+    }
+
+    #[derive(Deserialize)]
+    struct ParamsOwned {
+        name: String,
+    }
+
+    async fn handler(
+        req: &HttpRequest,
+        (data, params1): (Option<&web::Data<usize>>, web::Path<ParamsOwned>),
```

Comment:
why does `web::Path<Params<'_>>` not work?

Reply:
yeah, this is  known limitation.  See https://github.com/actix/actix-web/pull/2452#issuecomment-974599063